### DISCUSSION TITLE
jp2k: libvips built without JPEG2000 support

### DISCUSSION THREAD

**Discussion by petoor** [libvips/libvips#2366] [Q&A]
Hi John
Trying to save a jpeg 2000 image with tiffsave i get
```
pyvips.error.Error: unable to call tiffsave
jp2k: libvips built without JPEG2000 support

```
Currently i build libvips with libjpeg-turbo8-dev, which package am i missing?
Also is there a better choice than libjpeg-turbo8-dev?

**Comment by jcupitt** [comment]
Hi @petoor, you need OpenJPEG for jp2k support. I think the package is `libopenjp2-7-dev`.

**Reply by petoor** [reply]
Hmm. So far adding it prior to building libvips doesnt work.

I build libvips in a docker file.  It looks like this.

RUN apt install -y libpng-dev libtiff5-dev liborc-0.4-dev libwebp-dev libnifti-dev libopenslide-dev libccfits-dev libexif-dev libjpeg-turbo8-dev libgif-dev imagemagick libmagickcore-dev libopenjp2-7-dev
RUN apt install -y libmagick++-6.q16hdri-dev

WORKDIR /usr/src/dependencies
RUN git clone --branch v8.11.2 https://github.com/libvips/libvips.git
RUN cd libvips && ./autogen.sh && make -j$(($(nproc) - $overhead)) && make install 
ENV VIPSHOME=/usr/local
ENV LD_LIBRARY_PATH=$LD_LIBRARY_PATH:$VIPSHOME/lib

If i exec into the docker it tells me that both libopenjp2-7-dev and libopenjp2-7 is installed. However i still get the libvips built without JPEG2000 support

Any idea?

**Reply by jcupitt** [reply]
I think your openjpeg is probably too old. libvips needs openjpeg 2.4 or later -- the 2-7 in the debian name is not the openjpeg version number (confusingly). Have a look at the table at the end of libvips configure.

Debian bullseye has a version that's new enough, and I guess ubuntu 21.10 will too. If you are using an older platform, you'll need to build openjpeg from source, or find a backported .deb.

(openjpeg 2.3 is the version you'll come across most often, but it can't handle chroma subsampling and has wonky threading support, so libvips does not work with it)

**Reply by petoor** [reply]
Alright, ill try that.
It is not urgent that i add jpeg2000 support just now so it will have to wait a bit.

Do you plan on adding jpeg-xl compression to tiff as well in the future?

**Reply by jcupitt** [reply]
I don't think anyone is using jxl in tiff yet (as far as I know?) so I think I'll wait.

You're right that it looks very promising and no doubt use will spread.

**Reply by petoor** [reply]
If it does the same job as jp2k and jpeg but better and faster, i'd love to test it.

**Reply by jcupitt** [reply]
JXL is supported already, perhaps you saw?

```
$ vips copy k2.jpg k2.jxl
$ ls -l k2.jpg k2.jxl
-rw------- 1 john john 271558 Nov  2  2020 k2.jpg
-rw-rw-r-- 1 john john 175354 Jul 29 17:10 k2.jxl
```

It's faster than HEIC and AVIF.

**Reply by petoor** [reply]
Oh yeah i saw, but to use jxl inside tiffsave as compression.
For various reason i need to save my images as tiff.

**Reply by petoor** [reply]
Also i need JXL to support pyramids in order to be useful in the tiffsave (for my needs).

**Reply by jcupitt** [reply]
Ah OK. 

It'll probably be a couple of years yet, but yes, I imagine official libtiff JXL-in-TIFF will come.

**Reply by petoor** [reply]
Great! Ill set a reminder for 2025 😄
